### ISSUE TITLE
build bios.S with cross toolchain

### ISSUE THREAD

**Issue by stsp** [stsp/dosemu2#732]
Currently bios.S can only be built with the
native tool-chain, as it then gets linked in.
This is a big portability problem.
Even on macos-x, which uses x86 hw, the
support for 16bit x86 asm is very buggy, #721.
Not to speak about arm hw or emscripten,
where x86 asm doesn't exist at all.
We need a way to link in bios that is built
with cross tool-chain.

**Comment by stsp** [comment]
Maybe we can, but unless I am missing
something, this doesn't change anything.
I don't think the i386 object will link with
arm objects, no matter was it nasm or gas.

**Comment by stsp** [comment]
Actually HJLu "fixed" gas, not the linker...
The linker keeps producing the corrupted
executables from old object files, so I
don't know what he is up to.

**Comment by stsp** [comment]
We reference it, thats the problem.
Before thinking how to include the binary,
we need to avoid using its symbols
directly. The easiest way seem to be
to define its symbols to their offsets,
because this is what you already done
for dosdebug. We just need it in a
form of multiple defines rather than
an array.

**Comment by andrewbird** [comment]
Sure I was assuming we'd calculate the offsets as you suggested, just considering how to avoid linking with the dosemu binary.

**Comment by stsp** [comment]
Link it statically, convert to binary with
objcopy, and translate into a header file
with something like xxd I suppose.
We do the same steps for dos binaries,
except the translation to header.

**Comment by andrewbird** [comment]
Okay, sounds good.
I generated the offset header file, at least for now I prefixed the symbols with 'bios_offset_' to avoid a clash. See https://github.com/andrewbird/dosemu2/commit/86d2be3667ebbbc21d5cde028c64bc260866cddb but for it to be included from anywhere we need to tell make that it must be made first before any other file. How best to accomplish that in Dosemu build process?

~~~
// Warning: autogenerated                                                       
                                                                                
#define bios_offset_bios_f000 0x0000                                            
#define bios_offset_DPMI_dummy_start 0xc800                                     
#define bios_offset_DPMI_dpmi_init 0xc808                                       
#define bios_offset_DPMI_return_from_dos 0xc809                                 
#define bios_offset_DPMI_return_from_rmint 0xc80a                               
#define bios_offset_DPMI_return_from_realmode 0xc80b                            
#define bios_offset_DPMI_return_from_dos_memory 0xc80c                          
#define bios_offset_DPMI_int1c 0xc80d                                           
#define bios_offset_DPMI_int23 0xc80f                                           
#define bios_offset_DPMI_int24 0xc811                                           
~~~

**Comment by andrewbird** [comment]
Next question: Bios.S uses some of the symbols itself for instance
```asm
        lcall   *%cs:bios_f000_int10ptr-bios_f000
```
Can that be resolved at assembly time, isn't bios_f000 == 0?


```asm
L10:    /* chain to original handler (probably the video bios) */               
        ljmp    *%cs:bios_f000_int10_old-0x8000-bios_f000                       
```
And this one I don't get at all, what's the magic 0x8000?

**Comment by stsp** [comment]
Why not to include the header from
bios.h? It provides the same set of
symbols.

8000 is ROMBIOSSEG.
bios_foo is not 0 when bios.o is
linked into an executable. But after
your changes, it will indeed be 0.

**Comment by stsp** [comment]
I think we better keep subtracting
bios_foo so that all generated offsets
are relative.

**Comment by stsp** [comment]
i.e. asm doesnt know that bios_foo is 0.

**Comment by andrewbird** [comment]
Okay so in the case of `bios_f000_int10ptr - bios_f000` that's finding the offset by subtracting the final global values of the labels. If those labels were local would bios_f000 be 0?

**Comment by stsp** [comment]
You can use local labels directly, if you use short jumps.
In this case asm will try to calculate the offset between
the current position and the label - this is because short
jumps use relative offsets.
But if you use long jump, IIRC it doesn't have the relative
form in x86. I think in theory it is possible to convince asm
to not generate the relocation, but the local label is not
working this way:
```
--- a/src/base/bios/bios.S
+++ b/src/base/bios/bios.S
@@ -362,8 +362,8 @@ L10:        /* chain to original handler (probably the video bios) */
        jmp 30f         // EB xx for compat
        .space 7        // padding
        30: lret
-       1: ljmp $BIOSSEG,$int33_cont-bios_f000
-       int33_cont:
+       1: ljmp $BIOSSEG,$111f
+       111:
        ljmp *%cs:int33_chain-bios_f000
 
 /* ----------------------------------------------------------------- */
```

This gives:
```
/usr/bin/ld: lib/libbase_bios.a(bios.o): relocation R_X86_64_16 against `.text' can not be used when making объект PIE;
```

But subtract bios_foo and it works.

**Comment by andrewbird** [comment]
Yes thanks, will do. Fighting with `make` at the moment.

**Comment by stsp** [comment]
Note that when you link it statically, subtracting bios_foo
will not be needed, but you will need to use --section-start
to set the start address to 0. But if you subtract bios_foo,
you'll likely get the code w/o relocations, so --section-start
and friends will not be needed.
I think we can even do the same trick for other DOS progs
we compile, like generic.com, isemu.com, .sys files etc.
We would need to subtract the start symbol when accessing
data, and the relocations are avoided. For .com files you'd
also need to add 0x100 to compensate for psp.
So this is similar to what linker does.
We can eventually decide if we want to do the relocations
by hands, as in bios.S, or ask linker to do that for us,
specifying --section-start. If we decide what is better,
we can bring all places to the same technique. So if you
decide to remove bios_foo and set the start address to 0,
this question will be closed, as I leave it to the implementer.

**Comment by mrfixit2001** [comment]
It doesn't appear this is buildable using a cross-compiler for 32-bit ARM anymore. There's currently some hard-coded "--32" which the cross-compiler doesn't support, and with them removed it then results in tons of "bad instruction" and "ARM register expected" errors.

**Comment by stsp** [comment]
Should now be fixed.

**Comment by mrfixit2001** [comment]
@stsp thank you for the quick reply!! 
It looks like compiling the bios assembly still tries to use --32, here's the full build failure.

/output/host/bin/arm-buildroot-linux-gnueabihf-cpp -imacros config.hh -MD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -Wno-unused-result -Wcast-qual -Wwrite-strings -Wno-address-of-packed-member -fplan9-extensions -fno-pie   -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64  -O3 -g0 -DNDEBUG  -I/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/include -I/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/include/SDL2 -I/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/include/SDL" -I../../../../src/include -I../../../../src/plugin/include -I../../../../src/base/bios/x86 -I/output/build/dosemu2-d052e31f275ab29c594382390dc21fb6a52e10fe/src/include -I/output/build/dosemu2-d052e31f275ab29c594382390dc21fb6a52e10fe/src/base/lib -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -I/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/include -I/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/include/SDL2 -I/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/include/SDL -DBIOS_START=0xe000 -x assembler-with-cpp /output/build/dosemu2-d052e31f275ab29c594382390dc21fb6a52e10fe/src/base/bios/x86/bios.s | /output/host/bin/arm-buildroot-linux-gnueabihf-as --32 -o bios2.o
bdftopcf /output/build/dosemu2-d052e31f275ab29c594382390dc21fb6a52e10fe/etc/vga8x19.bdf | gzip -c -9 -n > ../etc/Xfonts/vga8x19.pcf.gz
/output/host/bin/arm-buildroot-linux-gnueabihf-as: unrecognized option '--32'
make[5]: *** [Makefile:17: bios0.o] Error 1
/output/build/dosemu2-d052e31f275ab29c594382390dc21fb6a52e10fe/src/Makefile.common.post:67: ../src/base/bios/x86/bios_data.xxd] Error 2
make[5]: *** [Makefile:17: bios2.o] Error 1
make[4]: *** /output/build/dosemu2-d052e31f275ab29c594382390dc21fb6a52e10fe/src/Makefile.common.post:67: ../src/base/bios/x86/bios_offsets.hh] Error 2
make[3]: *** [Makefile:29: default] Error 2

**Comment by stsp** [comment]
Please show configure output.
It looks for `i386-elf-as` which was
probably not found on your system,
so it resorted to native as and --32.

**Comment by stsp** [comment]
```
checking for as... /output/host/bin/arm-buildroot-linux-gnueabihf-as
checking for i386-elf-as... no
```

So... what's the name of your
cross-assembler?

**Comment by mrfixit2001** [comment]
The first one should be correct, it's /output/host/bin/arm-buildroot-linux-gnueabihf-as

**Comment by stsp** [comment]
First one is not cross-assembler.
Its a native arm assembler.
What's the name of the cross
assembler?

**Comment by stsp** [comment]
Also we already have the buildroot
setup. Please check out our fdpp-buildroot
git. I don't think you should try that
on your own, as its quite difficult.
Use our buildroot as a start.

**Comment by stsp** [comment]
So you need either an x86_64
cross-assembler, in which case
--32 is needed. Or you need the
i386 cross assember, in which
casae --32 will not be used.
But arm assembler is out of any
help here.

**Comment by mrfixit2001** [comment]
@stsp thank you, awesome details all around. Ok, so I have this to allow compiling my 64-bit kernel (in my 32-bit userland):
  /output/host/lib/gcc-linaro-aarch64-linux-gnu/bin/aarch64-linux-gnu-as
But thus far it's not used for anything other than my kernel. 

Looking over the fdpp-buildroot repo, I see your package is setup to depend on BR2_i386 or BR2_x86_64, but I'm not using either, I'm building BR2_arm. Maybe this just "won't work" then?

**Comment by stsp** [comment]
Buildroot will work in qemu, its not a
problem. You need to set up it as x86_64,
it will build all the needed cross compilers
and dosemu2, and you'll be able to run
it under qemu.
If you wanted the native build, why have
you used buildroot?
Yes, native arm build is impossible.

**Comment by mrfixit2001** [comment]
I am building linux distributions based on buildroot for ARM single-board-computers similar to a raspberry pi. 
It's not a buildroot issue so much as a cross-compiler incompatibility. I have multiple forks of dosbox compiled and running fine, so I was hoping to try dosemu2 as well for speed comparison. Thanks for your time and efforts!

**Comment by stsp** [comment]
dosemu2 takes advantage of kvm, which
on arm would not be useful for running
x86 code. So we can assume dosbox
to win that battle anyway. :)

**Comment by stsp** [comment]
Hi @mrfixit2001 !
We now have an arm builds on PPA.
I can try to help you with buildroot, too,
as I am trying to bring an arm support
in a somewhat "valuable" state.
Are you still interested in that experiment
of yours?
If so - you are welcome to open a new
ticket or a discussion thread.

**Comment by stsp** [comment]
Or for example please join topic #1788
